collect_app/src/main/java/org/odk/collect/android/logic/FormDetails.java

```diff
@@ -87,4 +90,48 @@ public boolean isNewerFormVersionAvailable() {
     public boolean areNewerMediaFilesAvailable() {
         return areNewerMediaFilesAvailable;
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        FormDetails that = (FormDetails) o;
+        return isNewerFormVersionAvailable() == that.isNewerFormVersionAvailable() &&
+                areNewerMediaFilesAvailable == that.areNewerMediaFilesAvailable &&
+                Objects.equals(getErrorStr(), that.getErrorStr()) &&
+                Objects.equals(getFormName(), that.getFormName()) &&
+                Objects.equals(getDownloadUrl(), that.getDownloadUrl()) &&
+                Objects.equals(getManifestUrl(), that.getManifestUrl()) &&
+                Objects.equals(formID, that.formID) &&
+                Objects.equals(getFormVersion(), that.getFormVersion()) &&
+                Objects.equals(getHash(), that.getHash()) &&
+                Objects.equals(getManifestFileHash(), that.getManifestFileHash());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getErrorStr(), getFormName(), getDownloadUrl(), getManifestUrl(), formID, getFormVersion(), getHash(), getManifestFileHash(), isNewerFormVersionAvailable(), areNewerMediaFilesAvailable);
+    }
+
+    public static FormDetails toFormDetails(FormListItem formListItem) {
+        return toFormDetails(formListItem, null, false, false);
+    }
+
+    public static FormDetails toFormDetails(FormListItem formListItem, String manifestFileHash, boolean isNewerFormVersionAvailable, boolean areNewerMediaFilesAvailable) {
```

Comment:
I'm not loving all this. Isn't `FormDetails` a wrapper for a `FormListItem` that provides status information? Maybe a `FormDetails` should have a `FormListItem` field?

Reply:
Yeah actually this is something I meant to come back to. Initially I'd thought I'd be getting rid of `FormDetails` so this was a way of converting between those values while I worked towards that. You're right that it seems really funky now. I'll take a look at cleaning it up.